### PR TITLE
Separate extern functions

### PR DESCRIPTION
**Pull request by talekhinezh**

### CODE REVIEW

File: radix-engine/src/wasm/prepare.rs [hunk 1]
```diff
@@ -161,6 +163,32 @@ impl WasmModule {
...
+                    x.field().eq(func_name) && matches!(x.internal(), Internal::Function(_))
+                }) {
+                    return Err(PrepareError::MissingExport {
+                        export_name: func_name.to_string(),
+                    });
+                }
+            }
+        }
+
+        Ok(self)
```
Comment (iamyulong) [inline radix-engine/src/wasm/prepare.rs:189]: Should we reject exports that's not specified in the ABI? :)

File: radix-engine/tests/abi/src/lib.rs [hunk 2]
```diff
@@ -25,12 +25,12 @@ blueprint! {
 }
 
 #[no_mangle]
-pub extern "C" fn AbiComponent2_main(_input: *mut u8, _input2: *mut u8) -> *mut u8 {
+pub extern "C" fn AbiComponent2_main(_input: *mut u8) -> *mut u8 {
```
Comment (iamyulong) [inline radix-engine/tests/abi/src/lib.rs:28]: Still `Blueprint_main`? Are we supporting a hybrid mode?

Comment (iamyulong) [inline radix-engine/tests/abi/src/lib.rs:28]: Still `Blueprint_main`?

File: scrypto-derive/src/blueprint.rs [hunk 3]
```diff
@@ -278,23 +243,28 @@ fn generate_dispatcher(bp_ident: &Ident, items: &[ImplItem]) -> Result<(Vec<Expr
...
-                    attrs: vec![],
-                    label: None,
-                    block: Block {
-                        brace_token: Brace {
-                            span: Span::call_site(),
-                        },
-                        stmts,
-                    },
-                }));
+                let function_ident = format_ident!("{}_{}_main", bp_ident, ident);
```
Comment (iamyulong) [inline scrypto-derive/src/blueprint.rs:248]: Should we get rid of the `_main` suffix?

Comment (iamyulong) [inline scrypto-derive/src/blueprint.rs:248]: Should we get rid of the `_main` suffix?

File: radix-engine/tests/abi/src/lib.rs (hunk 2 above)
Comment (talekhinezh) [reply radix-engine/tests/abi/src/lib.rs:28]: ah, it's just another level of indirection, ie every `fn_ident` must map to some extern function (in this case it is to the same function).

File: scrypto-derive/src/blueprint.rs (hunk 3 above)
Comment (talekhinezh) [reply scrypto-derive/src/blueprint.rs:248]: That's a good idea

File: radix-engine/src/wasm/prepare.rs (hunk 1 above)
Comment (talekhinezh) [reply radix-engine/src/wasm/prepare.rs:189]: hrm do you think that gives us more security? There may be interesting usecases if we allowed other exports?

Comment (iamyulong) [reply radix-engine/src/wasm/prepare.rs:189]: Not really, just more control over the structure of the wasm. Let's keep it as-in for now.

**Review by iamyulong — APPROVED** [review]

**Pull request merged**